internal/impl/etcd/util.go

```diff
@@ -0,0 +1,35 @@
+package etcd
+
+import (
+	"unicode/utf8"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+func etcdEventsToMap(events []*clientv3.Event) []map[string]any {
+	eventsMap := make([]map[string]any, len(events))
+
+	for _, e := range events {
+		event := map[string]any{
+			"key":             e.Kv.Key,
+			"value":           e.Kv.Value,
+			"type":            string(e.Type),
+			"version":         e.Kv.Version,
+			"mod_revision":    e.Kv.ModRevision,
+			"create_revision": e.Kv.CreateRevision,
+			"lease":           e.Kv.Lease,
+		}
+
+		if utf8.Valid(e.Kv.Key) {
+			event["key"] = string(e.Kv.Key)
+		}
+
+		if utf8.Valid(e.Kv.Value) {
+			event["value"] = string(e.Kv.Value)
+		}
+
+		eventsMap = append(eventsMap, event)
+	}
+
+	return eventsMap
+}
```

Comment:
```suggestion
package etcd

import (
	"unicode/utf8"

	mvccpb "go.etcd.io/etcd/api/v3/mvccpb"
	clientv3 "go.etcd.io/etcd/client/v3"
)

func etcdEventsToMap(events []*clientv3.Event) []map[string]any {

	eventsMap := make([]map[string]any, len(events))

	for i, e := range events {
		event := map[string]any{
			"key":             e.Kv.Key,
			"value":           e.Kv.Value,
			"type":            mvccpb.Event_EventType_name[int32(e.Type)],
			"version":         e.Kv.Version,
			"mod_revision":    e.Kv.ModRevision,
			"create_revision": e.Kv.CreateRevision,
			"lease":           e.Kv.Lease,
		}

		if utf8.Valid(e.Kv.Key) {
			event["key"] = string(e.Kv.Key)
		}

		if utf8.Valid(e.Kv.Value) {
			event["value"] = string(e.Kv.Value)
		}

		eventsMap[i] = event
	}

	return eventsMap
}
```

Couple of things changed: 

- don't append because you will get [null, {...}] 
- the value of the type needs to be resolved

Reply:
also not sure about the need to put them into an array - in what cases would you get more than 1 event in an array? (not familar with etcd)

Reply:
>  - don't append because you will get [null, {...}]

Oh wow that was a massive oversight on my end. Nice catch.

> - the value of the type needs to be resolved

The underlying event type has a `String()` method defined so I'm not sure the extra step is needed:
```golang
func (x Event_EventType) String() string {
	return proto.EnumName(Event_EventType_name, int32(x))
}
```

> also not sure about the need to put them into an array - in what cases would you get more than 1 event in an array? (not familar with etcd)

If I were watching a prefix `/foo` which contained `/baz` `/buz` and `/foobar`, and I deleted `/foo` (thereby also removing `/baz` `/buz` and `/foobar`) I believe this would register as multiple delete events for each key. Let me extend the integration tests to see if this actually is the case.